### ISSUE TITLE
Issue with user auth redirects

### ISSUE THREAD

**Issue by segheysens** [ga-wdi-exercises/project2#102]
Hi Instructors! I'm running into a conundrum with my user auth. I'm using devise, but I'm pretty sure that's not going to be the root of the issue.

I want to ensure that a user is authenticated prior to being able to use the rest of the site. My root is set to a static page (I have static_page_controller & views/static_pages), but I want it to automatically redirect to my /accounts/sign_in (devise/sessions#new) page if the user hasn't already signed in. I think it's just a matter of finding the correct placement of my _before_filter :authenticate_user!_ method. Right now, when I go to the root (static_pages#index) I'm receiving an error that the @user variable is nil, I realize that this is probably because the root helps skip the _before_filter :authenticate_user!_ at the top of the controller. (This'll probably make more sense if you take a look at the code below.)

config/routes.rb:

``` ruby
Rails.application.routes.draw do
  devise_for :users, :path => 'accounts'
  # The priority is based upon order of creation: first created -> highest priority.
  # See how all your routes lay out with "rake routes".

  # You can have the root of your site routed with "root"
  root 'static_pages#index'

  # get "/index" => "static_pages#index"
  get "/submit" => "static_pages#submit"
  get "/search" => "static_pages#search"

  resources :users, only: [:index, :show ] do
    resources :products
  end
end
```

app/controllers/application_controller.rb:

``` ruby
class ApplicationController < ActionController::Base
  # Prevent CSRF attacks by raising an exception.
  # For APIs, you may want to use :null_session instead.
  protect_from_forgery with: :exception

  #before_filter :authenticate_user!

  protected
  def authenticate_user!
   if user_signed_in?
     super
   else
     redirect_to new_user_session_path
   end
 end

end
```

app/controllers/static_pages_controller.rb:

``` ruby
class StaticPagesController < ApplicationController

  before_filter :authenticate_user!

  def index
    @user = User.find(session[:user]["id"])
  end

  def search
    @user = User.find(session[:user]["id"])
  end

  def submit
    @user = User.find(session[:user]["id"])
  end

end
```

The relevant view - app/views/static_pages/index.html

``` ruby
<h1> Welcome to Bewze!</h1>

<h2><a href="/search" class=".searchsubmit">Search</a></h2>

<% if user_signed_in? %>
  <h2><a href="<%= new_user_product(@user.id) %>" class=".searchsubmit">Submit</a></h2>
<% end %>
```

tl;dr - I need to figure out where to user-authenticate without creating an redirect loop, but guaranteeing users log in before being able to do anything. There's probably a simple solution to this that I haven't thought of, thanks for taking the time to review this!

**Comment by andrewsunglaekim** [comment]
have you tried using a `before_action :authenticate_user!` in the ApplicationController?

**Comment by jshawl** [comment]
Andy's comment should fix this! Check out the documentation as well. https://github.com/plataformatec/devise#controller-filters-and-helpers

Going to close but please reopen if this doesn't fix!

**Comment by segheysens** [comment]
So I tried moving the _before_action authenticate_user!_ to my application controller, and it didn't resolve the issue. I think it's because I'm rendering that index file as my root, and it's skipping the application controller? (Unless the application controller runs no matter what, I can't find a clear answer on that...)

Also, I tried making this change in my StaticPagesController to no eval (it's still throwing the "no method for nil" error in reference to @user:

``` ruby
class StaticPagesController < ApplicationController

  def index
    if user_signed_in?
      @user = User.find(session[:user]["id"])
    end
  end

```

I'll be heading in shortly (about 2pm) to work on this, maybe some side-by-side pair programming will do the trick.
